# Predicting Average House Prices in Madrid Neighbourhoods Using Socioeconomic Indicators

## Import Libraries

In [2]:
# Import the Pandas library for data manipulation and analysis
import pandas as pd

## Load Datasets

In [3]:
# Loading Dataset 2: Average Declared Housing Price (euros/m2) by District and Neighborhood in Madrid
prices = pd.read_csv("../data/prices.csv", sep=";",skiprows=5)

## 2. Data Cleaning

### Observations 
Based on the initial exploration of the dataset, the following data cleaning and preprocessing steps will be carried out on a working copy of the dataset:

- Combine the header and the first two rows to create meaningful column names.
- Remove the final four rows containing metadata.
- Rename columns to provide descriptive English labels.
- Convert relevant columns to the appropriate numerical data types and reformat values to standard decimal notation.
- Remove rows representing the City of Madrid and district-level data, retaining only neighbourhood-level observations.
- Filter the dataset to include only 2025 data.

In addition, the neighbourhood names will be examined and standardised where necessary to facilitate merging with the socioeconomic indicators dataset. This will ensure that a common key is available for joining the two datasets.

In [4]:
# Create a working copy of dataset
prices_clean = prices.copy()

In [5]:
# Display the first ten rows
prices_clean.head(10)

,Unnamed: 0,Unnamed: 1,2007,2007.1,2007.2,2007.3,2007.4,2007.5,2007.6,2007.7,...,2024.8,2025,2025.1,2025.2,2025.3,2025.4,2025.5,2025.6,2025.7,2025.8
0,NaN,NaN,Euros/m2,Euros/m2,Euros/m2,Números índice (Madrid=100),Números índice (Madrid=100),Números índice (Madrid=100),Números índice (Total=100),Números índice (Total=100),...,Números índice (Total=100),Euros/m2,Euros/m2,Euros/m2,Números índice (Madrid=100),Números índice (Madrid=100),Números índice (Madrid=100),Números índice (Total=100),Números índice (Total=100),Números índice (Total=100)
1,NaN,NaN,Total,Nuevas,Usadas,Total,Nuevas,Usadas,Total,Nuevas,...,Usadas,Total,Nuevas,Usadas,Total,Nuevas,Usadas,Total,Nuevas,Usadas
2,Ciudad de Madrid,Ciudad de Madrid,"3.746,27","3.479,98","3.859,89",100,100,100,100,"92,89",...,"100,39","5.285,72","5.041,46","5.333,59",100,100,100,100,"95,38","100,91"
3,01. Centro,01. Centro,"4.798,92","4.988,4","4.768,62","128,1","143,35","123,54",100,"103,95",...,"100,18","7.147,7","6.814,13","7.171,36","135,23","135,16","134,46",100,"95,33","100,33"
4,01. Centro,011. Palacio,..,..,..,..,..,..,..,..,...,"99,32","6.895,15","6.421,28","6.954,58","130,45","127,37","130,39",100,"93,13","100,86"
5,01. Centro,012. Embajadores,..,..,..,..,..,..,..,..,...,"99,95","6.054,75","6.153,41","6.044,57","114,55","122,06","113,33",100,"101,63","99,83"
6,01. Centro,013. Cortes,..,..,..,..,..,..,..,..,...,"99,71","7.246,75",0,"7.246,75","137,1",0,"135,87",100,0,100
7,01. Centro,014. Justicia,..,..,..,..,..,..,..,..,...,100,"9.541,9","9.505,61","9.544,48","180,52","188,55","178,95",100,"99,62","100,03"
8,01. Centro,015. Universidad,..,..,..,..,..,..,..,..,...,"99,6","7.135,43","6.682,45","7.149,96","134,99","132,55","134,06",100,"93,65","100,2"
9,01. Centro,016. Sol,..,..,..,..,..,..,..,..,...,"101,95","7.696,61",0,"7.726,49","145,61",0,"144,86",100,0,"100,39"


#### Combine the header and the first two rows to create appropriate column names

In [6]:
# Combine first two rows into header
new_columns = ( prices_clean.columns.astype(str) + "_" +
                prices_clean.iloc[1].fillna("").astype(str)+ ":_" +
                prices_clean.iloc[0].fillna("").astype(str) )

prices_clean.columns= new_columns

In [7]:
# Drop first two rows by resetting the indexes
prices_clean = prices_clean.iloc[2:].reset_index(drop=True)

In [8]:
# Display the first ten rows
prices_clean.head(10)

,Unnamed: 0_:_,Unnamed: 1_:_,2007_Total:_Euros/m2,2007.1_Nuevas:_Euros/m2,2007.2_Usadas:_Euros/m2,2007.3_Total:_Números índice (Madrid=100),2007.4_Nuevas:_Números índice (Madrid=100),2007.5_Usadas:_Números índice (Madrid=100),2007.6_Total:_Números índice (Total=100),2007.7_Nuevas:_Números índice (Total=100),...,2024.8_Usadas:_Números índice (Total=100),2025_Total:_Euros/m2,2025.1_Nuevas:_Euros/m2,2025.2_Usadas:_Euros/m2,2025.3_Total:_Números índice (Madrid=100),2025.4_Nuevas:_Números índice (Madrid=100),2025.5_Usadas:_Números índice (Madrid=100),2025.6_Total:_Números índice (Total=100),2025.7_Nuevas:_Números índice (Total=100),2025.8_Usadas:_Números índice (Total=100)
0,Ciudad de Madrid,Ciudad de Madrid,"3.746,27","3.479,98","3.859,89",100,100,100,100,"92,89",...,"100,39","5.285,72","5.041,46","5.333,59",100,100,100,100,"95,38","100,91"
1,01. Centro,01. Centro,"4.798,92","4.988,4","4.768,62","128,1","143,35","123,54",100,"103,95",...,"100,18","7.147,7","6.814,13","7.171,36","135,23","135,16","134,46",100,"95,33","100,33"
2,01. Centro,011. Palacio,..,..,..,..,..,..,..,..,...,"99,32","6.895,15","6.421,28","6.954,58","130,45","127,37","130,39",100,"93,13","100,86"
3,01. Centro,012. Embajadores,..,..,..,..,..,..,..,..,...,"99,95","6.054,75","6.153,41","6.044,57","114,55","122,06","113,33",100,"101,63","99,83"
4,01. Centro,013. Cortes,..,..,..,..,..,..,..,..,...,"99,71","7.246,75",0,"7.246,75","137,1",0,"135,87",100,0,100
5,01. Centro,014. Justicia,..,..,..,..,..,..,..,..,...,100,"9.541,9","9.505,61","9.544,48","180,52","188,55","178,95",100,"99,62","100,03"
6,01. Centro,015. Universidad,..,..,..,..,..,..,..,..,...,"99,6","7.135,43","6.682,45","7.149,96","134,99","132,55","134,06",100,"93,65","100,2"
7,01. Centro,016. Sol,..,..,..,..,..,..,..,..,...,"101,95","7.696,61",0,"7.726,49","145,61",0,"144,86",100,0,"100,39"
8,02. Arganzuela,02. Arganzuela,"3.770,95","3.472,37","3.825,39","100,66","99,78","99,11",100,"92,08",...,"89,47","6.086,21","6.776,24","5.648,69","115,14","134,41","105,91",100,"111,34","92,81"
9,02. Arganzuela,021. Imperial,..,..,..,..,..,..,..,..,...,"83,54","5.855,12","6.010,51","5.776,45","110,77","119,22","108,3",100,"102,65","98,66"


#### Filter the dataset to include only 2025 data

In [9]:
# Inspect the last 9 columns containing 2025 data

prices_clean.columns[-9:]

Index(['2025_Total:_Euros/m2', '2025.1_Nuevas:_Euros/m2',
       '2025.2_Usadas:_Euros/m2', '2025.3_Total:_Números índice (Madrid=100)',
       '2025.4_Nuevas:_Números índice (Madrid=100)',
       '2025.5_Usadas:_Números índice (Madrid=100)',
       '2025.6_Total:_Números índice (Total=100)',
       '2025.7_Nuevas:_Números índice (Total=100)',
       '2025.8_Usadas:_Números índice (Total=100)'],
      dtype='str')

In [10]:
# Create a new DF with 2025 prices data
prices_clean_2025_unformatted= prices_clean[["Unnamed: 0_:_","Unnamed: 1_:_",'2025_Total:_Euros/m2']].copy()

In [11]:
prices_clean_2025_unformatted.head(10)

,Unnamed: 0_:_,Unnamed: 1_:_,2025_Total:_Euros/m2
0,Ciudad de Madrid,Ciudad de Madrid,"5.285,72"
1,01. Centro,01. Centro,"7.147,7"
2,01. Centro,011. Palacio,"6.895,15"
3,01. Centro,012. Embajadores,"6.054,75"
4,01. Centro,013. Cortes,"7.246,75"
5,01. Centro,014. Justicia,"9.541,9"
6,01. Centro,015. Universidad,"7.135,43"
7,01. Centro,016. Sol,"7.696,61"
8,02. Arganzuela,02. Arganzuela,"6.086,21"
9,02. Arganzuela,021. Imperial,"5.855,12"


#### Rename columns to provide descriptive English labels.

In [12]:
# Rename columns 

column_mapper = {'Unnamed: 0_:_': 'district_name', 
                'Unnamed: 1_:_': 'neighbourhood_name',
                '2025_Total:_Euros/m2': 'mean_average_price_euros_per_m2',}

av_prices_clean_2025=prices_clean_2025_unformatted.rename(mapper=column_mapper, axis=1)

In [13]:
# Display DF 
av_prices_clean_2025.head()

,district_name,neighbourhood_name,mean_average_price_euros_per_m2
0,Ciudad de Madrid,Ciudad de Madrid,"5.285,72"
1,01. Centro,01. Centro,"7.147,7"
2,01. Centro,011. Palacio,"6.895,15"
3,01. Centro,012. Embajadores,"6.054,75"
4,01. Centro,013. Cortes,"7.246,75"


#### Remove the final four rows containing metadata and Remove rows representing the city of Madrid

In [14]:
# Count missing values
av_prices_clean_2025.isnull().sum()

district_name                      2
neighbourhood_name                 4
mean_average_price_euros_per_m2    4
dtype: int64

In [15]:
# Display rows with missing values 
av_prices_clean_2025[av_prices_clean_2025.isnull().any(axis=1)]

,district_name,neighbourhood_name,mean_average_price_euros_per_m2
153,NaN,NaN,NaN
154,NaN,NaN,NaN
155,Fuente: Colegio de Registradores de España,NaN,NaN
156,Observaciones: Tasa de cobertura: La informaci...,NaN,NaN


In [16]:
# Drop rows with missing values and 1st row as that is a city wide metric  
av_prices_clean_2025 = av_prices_clean_2025.drop([0,153,154,155,156])

In [17]:
# Display dimensions
av_prices_clean_2025.shape

(152, 3)

### Observations 

Following the removal of the metadata and the first row representing the City of Madrid, the dataset contains three columns representing the district name, neighbourhood name and average housing price. The dataset also contains 152 rows.

Madrid is divided into 21 official districts and 131 neighbourhoods. As 21 + 131 = 152, this confirms that the dataset now contains one observation for each district and neighbourhood, with only 2025 housing price data remaining.

In [18]:
# Display last 30 rows

av_prices_clean_2025.tail(30)

,district_name,neighbourhood_name,mean_average_price_euros_per_m2
123,17. Villaverde,17. Villaverde,"2.649,6"
124,17. Villaverde,"171. Villaverde Alto, Casco Histórico de Villa...","2.928,26"
125,17. Villaverde,172. San Cristóbal,"1.909,6"
126,17. Villaverde,173. Butarque,"3.143,01"
127,17. Villaverde,174. Los Rosales,"2.807,4"
128,17. Villaverde,175. Ángeles,"2.956,03"
129,18. Villa de Vallecas,18. Villa de Vallecas,3.278
130,18. Villa de Vallecas,181. Casco Histórico de Vallecas,"2.915,21"
131,18. Villa de Vallecas,182. Santa Eugenia,"2.899,99"
132,18. Villa de Vallecas,183. Ensanche de Vallecas,"3.552,31"


### Remove rows representing district-level data, retaining only neighbourhood-level data

In [19]:
# Display rows where district matches neighbourhood being distrect level observations
district_prices_2025= av_prices_clean_2025[
    av_prices_clean_2025["district_name"] == av_prices_clean_2025["neighbourhood_name"]
]

district_prices_2025

,district_name,neighbourhood_name,mean_average_price_euros_per_m2
1,01. Centro,01. Centro,"7.147,7"
8,02. Arganzuela,02. Arganzuela,"6.086,21"
16,03. Retiro,03. Retiro,"7.469,32"
23,04. Salamanca,04. Salamanca,"9.405,91"
30,05. Chamartín,05. Chamartín,"7.427,12"
37,06. Tetuán,06. Tetuán,"5.447,04"
44,07. Chamberí,07. Chamberí,"8.238,24"
51,08. Fuencarral-El Pardo,08. Fuencarral-El Pardo,"5.417,12"
60,09. Moncloa-Aravaca,09. Moncloa-Aravaca,"5.869,93"
68,10. Latina,10. Latina,"3.534,13"


In [20]:
# Display dimensions on districts

district_prices_2025.shape


(21, 3)

In [21]:
# Remove districts averages 

av_prices_clean_2025 = av_prices_clean_2025.loc[av_prices_clean_2025["district_name"] != av_prices_clean_2025["neighbourhood_name"]]


In [22]:
# Display dimensions

av_prices_clean_2025.shape

(131, 3)

#### Observations

Inspection of the rows where the district name and neighbourhood name are identical shows that these represent district-level observations rather than neighbourhood-level data. As the project focuses on neighbourhood-level analysis, these rows were removed from the dataset. This was confirmed using the shape attribute.

In [23]:
# Display DF 
av_prices_clean_2025.head()

,district_name,neighbourhood_name,mean_average_price_euros_per_m2
2,01. Centro,011. Palacio,"6.895,15"
3,01. Centro,012. Embajadores,"6.054,75"
4,01. Centro,013. Cortes,"7.246,75"
5,01. Centro,014. Justicia,"9.541,9"
6,01. Centro,015. Universidad,"7.135,43"


### Missing Values and Duplicates

In [24]:
# Count duplicate rows
av_prices_clean_2025.duplicated().sum()

np.int64(0)

In [25]:
# Count missing values
av_prices_clean_2025.isnull().sum()

district_name                      0
neighbourhood_name                 0
mean_average_price_euros_per_m2    0
dtype: int64

In [26]:
# Display data types
av_prices_clean_2025.dtypes

district_name                      str
neighbourhood_name                 str
mean_average_price_euros_per_m2    str
dtype: object

In [27]:
# Checking if there are any neighbourhoods with no value
av_prices_clean_2025[av_prices_clean_2025['mean_average_price_euros_per_m2']== '0']


,district_name,neighbourhood_name,mean_average_price_euros_per_m2
149,21. Barajas,212. Aeropuerto,0


In [28]:
# Dropping airport 
av_prices_clean_2025 = av_prices_clean_2025.loc[
    av_prices_clean_2025["mean_average_price_euros_per_m2"] != '0'
]

In [29]:
# Checking if there are any neighbourhoods with no value
av_prices_clean_2025[av_prices_clean_2025['mean_average_price_euros_per_m2']== '0']

,district_name,neighbourhood_name,mean_average_price_euros_per_m2


### Observations 

Checking for missing values and duplicate records confirmed that the dataset is complete and contains no duplicate observations. However, inspection of neighbourhoods with an average housing price of 0 identified the airport. As the airport is not a residential neighbourhood, this observation was removed to improve the integrity of the model, which aims to predict average residential housing prices across Madrid neighbourhoods.

In [30]:
# Display info on the DF
av_prices_clean_2025.info()

<class 'pandas.DataFrame'>
Index: 130 entries, 2 to 152
Data columns (total 3 columns):
 #   Column                           Non-Null Count  Dtype
---  ------                           --------------  -----
 0   district_name                    130 non-null    str  
 1   neighbourhood_name               130 non-null    str  
 2   mean_average_price_euros_per_m2  130 non-null    str  
dtypes: str(3)
memory usage: 9.0 KB


In [31]:
# Previwing the first 5 rows

av_prices_clean_2025.head()

,district_name,neighbourhood_name,mean_average_price_euros_per_m2
2,01. Centro,011. Palacio,"6.895,15"
3,01. Centro,012. Embajadores,"6.054,75"
4,01. Centro,013. Cortes,"7.246,75"
5,01. Centro,014. Justicia,"9.541,9"
6,01. Centro,015. Universidad,"7.135,43"


In [ ]:
#### Create a common key for merge the two datasets

In [32]:
# Return the number of distinct district names
av_prices_clean_2025['neighbourhood_name'].nunique()

130

In [33]:
# Obtain the distinct neighbourhood names in order to compare to the neighbourhood names in the  prices dataset
av_prices_clean_2025["neighbourhood_name"].unique()

<ArrowStringArray>
[                   '011. Palacio',                '012. Embajadores',
                     '013. Cortes',                   '014. Justicia',
                '015. Universidad',                        '016. Sol',
                   '021. Imperial',                    '022. Acacias',
                    '023. Chopera',                    '024. Legazpi',
 ...
                    '203. Amposta',                      '204. Arcos',
                      '205. Rosas',                      '206. Rejas',
                 '207. Canillejas',                '208. El Salvador',
           '211. Alameda de Osuna', '213. Casco Histórico de Barajas',
                      '214. Timón',                 '215. Corralejos']
Length: 130, dtype: str

In [34]:
# Format neighbourhood names to remove the number, the dot and space.
av_prices_clean_2025.loc[:, "neighbourhood_name"] = (
    av_prices_clean_2025["neighbourhood_name"]
    .str.replace(r"^\d+\.\s*", "", regex=True)
    .str.strip()
)

In [35]:
# View first 5 rows to confirm

av_prices_clean_2025.head()

,district_name,neighbourhood_name,mean_average_price_euros_per_m2
2,01. Centro,Palacio,"6.895,15"
3,01. Centro,Embajadores,"6.054,75"
4,01. Centro,Cortes,"7.246,75"
5,01. Centro,Justicia,"9.541,9"
6,01. Centro,Universidad,"7.135,43"


### Format values to standard decimal notation

In [36]:
# Format the values: remove the full stops and replace commas with decimal point

av_prices_clean_2025 = av_prices_clean_2025.replace(r"\.", "", regex=True)
av_prices_clean_2025= av_prices_clean_2025.replace(",", ".", regex=True)

In [37]:
# Create exportable DF only of neighbourhood and price value
average_price_2025_prepared= av_prices_clean_2025[["neighbourhood_name","mean_average_price_euros_per_m2"]]

In [38]:
# View first 10 rows

average_price_2025_prepared.head(10)

,neighbourhood_name,mean_average_price_euros_per_m2
2,Palacio,6895.15
3,Embajadores,6054.75
4,Cortes,7246.75
5,Justicia,9541.9
6,Universidad,7135.43
7,Sol,7696.61
9,Imperial,5855.12
10,Acacias,5736.45
11,Chopera,5473.88
12,Legazpi,7113.37


#### Convert values to a numerical data type

In [39]:
# Convert all values to float 

average_price_2025_prepared['mean_average_price_euros_per_m2'].astype(float)

2      6895.15
3      6054.75
4      7246.75
5      9541.90
6      7135.43
        ...   
146    4735.53
148    5052.01
150    3962.41
151    4418.93
152    4890.30
Name: mean_average_price_euros_per_m2, Length: 130, dtype: float64

In [40]:
# Confirm data types as float
average_price_2025_prepared.dtypes

neighbourhood_name                 str
mean_average_price_euros_per_m2    str
dtype: object

In [41]:
# Sort data alphabetically by neighbourhood name and reset index 

average_price_2025_prepared=average_price_2025_prepared.sort_values('neighbourhood_name')

average_price_2025_prepared=average_price_2025_prepared.reset_index(drop=True)

In [42]:
# View first 10 rows of cleaned dataset
average_price_2025_prepared.head(10)

,neighbourhood_name,mean_average_price_euros_per_m2
0,Abrantes,2912.9
1,Acacias,5736.45
2,Adelfas,5741.86
3,Alameda de Osuna,5052.01
4,Almagro,10795.41
5,Almenara,4987.49
6,Almendrales,4323.83
7,Aluche,3360.3
8,Amposta,2544.1
9,Apóstol Santiago,4162.46


In [43]:
# Check dimensions of cleaned dataset
# Madrid has 130 neighbourhoods without the airport

average_price_2025_prepared.shape

(130, 2)

In [44]:
# Confirm no mising values of cleaned dataset

average_price_2025_prepared.isna().sum()

neighbourhood_name                 0
mean_average_price_euros_per_m2    0
dtype: int64

In [45]:
# View description of cleaned dataset

average_price_2025_prepared.describe()

,neighbourhood_name,mean_average_price_euros_per_m2
count,130,130
unique,130,130
top,Abrantes,2912.9
freq,1,1


In [46]:
# View summary of cleaned dataset

average_price_2025_prepared.info()

<class 'pandas.DataFrame'>
RangeIndex: 130 entries, 0 to 129
Data columns (total 2 columns):
 #   Column                           Non-Null Count  Dtype
---  ------                           --------------  -----
 0   neighbourhood_name               130 non-null    str  
 1   mean_average_price_euros_per_m2  130 non-null    str  
dtypes: str(2)
memory usage: 4.4 KB


In [47]:
# Save prepared datset

average_price_2025_prepared.to_csv(
    "../data/prices_2025_prepared.csv",
    index=False
)

### Observations

After completing the data cleaning steps described earlier in the notebook, the summary statistics confirm that the dataset contains 130 neighbourhood observations with no missing values. The dataset is therefore considered ready for merging with the socioeconomic indicators dataset and is exported as a CSV file for use in the subsequent preprocessing stage.